[ESPnet2] Update ASR Recipes

b-flo opened the pull request:
- [x] vivos
- [x] yesno

@kamo-naoyuki I took the liberty to fix a typo and change a log message in `egs2/asr1/TEMPLATE/asr.sh`, and temporary fix a test in `egs2/asr1/TEMPLATE/asr.sh`. Also:
-  It seems that if a non-linguistic symbols table (with `<blank>` included) is not specified the scoring will be messed up (`<blank>` will be attached to the symbol at the right or treated as a word -> "< b l a n k >"). Thus, I added the nlsyms file creation during data preparation for now.
- If `--use_wordlm` is set to true and a word LM config is passed through `--lm_config`, the LM will be treated as a non-word LM and `vocabsize`-`word_vocabsize` difference will cause a dimension mismatch in beam search ([Here in first](https://github.com/espnet/espnet/blob/28ce90a17148afcb36f4e593966911b9c3a6230b/espnet/nets/beam_search.py#L290)). I'm aware that word LM is not supported yet and of the test in `espnet2/bin/asr_decode.py` for `word_lm_train_config` but I don't know if the behaviour is intended?

----
Results (excluding yesno):

**VIVOS (no LM) - CER**

|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|decode_testdecode_asr_model_eval.loss.best|760|32771|83.3|11.7|4.9|3.0|19.7|96.6|
|decode_train_devdecode_asr_model_eval.loss.best|100|4007|83.6|12.8|3.6|2.7|19.0|99.0|

**VIVOS (LM) - CER**

|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|decode_testdecode_lm_eval.loss.best_asr_model_eval.loss.best|760|32771|85.5|11.0|3.5|3.5|18.0|91.6|
|decode_train_devdecode_lm_eval.loss.best_asr_model_eval.loss.best|100|4007|85.1|12.2|2.6|2.8|17.7|96.0|

sw005320 commented:
This PR is related to #1497

kamo-naoyuki commented:
Thank you for working for epsnet2!

>     * It seems that if a non-linguistic symbols table (with `<blank>` included) is not specified the scoring will be messed up (`<blank>` will be attached to the symbol at the right or treated as a word -> "< b l a n k >"). 

In my understanding, `<blank>` shouldn't appear in the decoding result originally.
I'm not why it happens and there are some bugs.

>     * If `--use_wordlm` is set to true and a word LM config is passed through `--lm_config`, the LM will be treated as a non-word LM and `vocabsize`-`word_vocabsize` difference will cause a dimension mismatch in beam search ([Here in first](https://github.com/espnet/espnet/blob/28ce90a17148afcb36f4e593966911b9c3a6230b/espnet/nets/beam_search.py#L290)). I'm aware that word LM is not supported yet and of the test in `espnet2/bin/asr_decode.py` for `word_lm_train_config` but I don't know if the behaviour is intended?

Sorry, I should put any comments in asr.sh such that Word LM is not supported yet because decoding part doesn't support it. You can ignore word lm part entirely.

kamo-naoyuki commented on egs2/TEMPLATE/asr1/db.sh:
```diff
@@ -17,6 +17,8 @@ if [[ "$(hostname -d)" == clsp.jhu.edu ]]; then
                 ## Case merl :MERL setup. Neccesary directory is WAV and sdb
     LIBRISPEECH=
     VOXFORGE=
+    YESNO=
```
`db.`sh is intend to write the path to the external corpus or the default path to put the downloaded corpus.
YESNO is very small corpus, so you don't need to write it here.

kamo-naoyuki commented on egs2/vivos/asr1/run.sh:
```diff
@@ -0,0 +1,43 @@
+    --stop_stage ${stop_stage}                  \
+    --nj ${nj}                                  \
+    --decode_asr_model ${decode_asr_model}      \
+    --feats_type fbank_pitch                    \
```
Could you test with feats_type=raw? It indicates on-the-fly feature extraction using librosa.

Basically we shouldn't depend on Kaldi command as possible  and `raw` is standard method in epsnet2.
 `fbank_pitch` is manly put for reproducing the previous result and also as an example to use the extracted features from external system.

b-flo replied:
I also trained and tested with `feats_type=raw` however I used `fbank_pitch` as default to reproduce espnet1 results. Do you want me to change defaults and results?

@sw005320 @kamo-naoyuki Should I try to reproduce recipes closely or I should use new defaults and revisit (a bit) recipes?

kamo-naoyuki replied:
Thank you. Yes, it's better to change the default to feats_type=raw if there is not significant degradation. 

I think we do not always need to reproduce the result even if small degradation is observed.
Now is good time to revisit recipes if you have better idea.

b-flo replied:
OK! I'll update recipe/results after retraining, I erased exp folders for raw type yesterday..

b-flo replied:
| dataset | feats_type | CER |
| --- | --- | --- |
| dev (no LM) | fbank_pitch | 19.0 |
| dev (LM) | fbank_pitch | 17.7 |
| dev (no LM) | raw | 20.9 |
| dev (LM) | raw | 19.0 |
| test (no LM) | fbank_pitch | 19.7 |
| test (LM) | fbank_pitch | 18.0 |
| test (no LM) | raw | 20.8 |
| test (LM) | raw | 18.5 |

Is the degradation ok for you?

kamo-naoyuki replied:
hmm, what do you think @sw005320 ?

sw005320 replied:
I think it's OK, but I found that I saw a significant degradation in the WER without LM. I don't know what happened.

b-flo replied:
Yes I observed the same thing. However CER is on par with reported results in `egs/` and WER isn't reported so I didn't think there was a problem.

sw005320 replied:
But https://github.com/espnet/espnet/blob/49d471bfa359aff182c58a65a73cb63891215001/egs2/vivos/asr1/README.md#asr_train_fbank_pitch_char looks normal???

b-flo replied:
Thanks, I found the problem! This WER was related to the first bullet in my original post, I did a mistake when parsing folders. Updated!

kamo-naoyuki commented on egs2/vivos/asr1/local/data.sh:
```diff
@@ -0,0 +1,109 @@
+eval_set="test"
+
+if [ ${stage} -le -1 ] && [ ${stop_stage} -ge -1 ]; then
+    echo "stage -1: Data download"
```
Use `log` instead of `echo` for logging purpose

kamo-naoyuki commented on egs2/vivos/asr1/local/data.sh:
```diff
@@ -0,0 +1,109 @@
+    mkdir -p ${datadir}
+
+    if [ -d ${datadir}/vivos ] || [ -f ${datadir}/vivos.tar.gz ]; then
+        echo "$0: vivos directory or archive already exists in ${datadir}. Skipping download."
```
ditto

kamo-naoyuki commented on egs2/vivos/asr1/local/data.sh:
```diff
@@ -0,0 +1,109 @@
+        echo "$0: vivos directory or archive already exists in ${datadir}. Skipping download."
+    else
+        if ! which wget >/dev/null; then
+            echo "$0: wget is not installed."
```
ditto

b-flo commented:
> In my understanding, `<blank>` shouldn't appear in the decoding result originally.
> I'm not why it happens and there are some bugs.

One example of text produced (yesno) without providing nlsyms : 
  `0_0_0_0_1_1_1_1 <blank>NO NO NO NO NO NO NO`

And the corresponding hyp during scoring : 
  `HYP:  < B L A N K > N O <SPACE> ...`

I didn't try to pinpoint the problem as I roughly guessed it could be circumvented with nlsyms but I can go through related code to find the problem if you want!

> Sorry, I should put any comments in asr.sh such that Word LM is not supported yet because decoding part doesn't support it. You can ignore word lm part entirely.

Sorry I think I didn't explain correctly.
There is already a test for word LM : 
https://github.com/espnet/espnet/blob/4c174113fce19ad6551a6f70eb0f8a2371c5c903/espnet2/bin/asr_decode.py#L61-L62

However if you use the same "procedure" as in espnet1 and specify word LM config through `--lm_config` in `run.sh` with `use_wordlm=True`, it won't raise any error.

https://github.com/espnet/espnet/blob/4c174113fce19ad6551a6f70eb0f8a2371c5c903/egs2/TEMPLATE/asr1/asr.sh#L662-L665

Because `use_wordlm` is not checked in `asr.sh`, char/word LM will be treated the same and passed through `bin.asr_decode.py`

https://github.com/espnet/espnet/blob/4c174113fce19ad6551a6f70eb0f8a2371c5c903/egs2/TEMPLATE/asr1/asr.sh#L695-L703

kamo-naoyuki commented:
As far word LM, I fixed asr.sh in #1498 . Please check it.

> I didn't try to pinpoint the problem as I roughly guessed it could be circumvented with nlsyms but I can go through related code to find the problem if you want!

I just ran your branch of yesno for now without nlsyms, but I couldn't reproduce the `<blank>`.
Could you show me the `token_int` and `token` file? I'm not sure where the problem is. 
If you have time, would you investigate this problem?

kamo-naoyuki commented on egs2/vivos/asr1/local/data.sh:
```diff
@@ -0,0 +1,109 @@
+stage=-1
+stop_stage=1
+
+datadir=./downloads
```
Use `$VIVOS` here instead of `./downloads` and write it to db.sh.

kamo-naoyuki commented on egs2/yesno/asr1/local/data.sh:
```diff
@@ -0,0 +1,118 @@
+stage=-1
+stop_stage=1
+
+datadir=./downloads
```
Sorry for again, I changed my mind, so could you also use ${YESNO} here and write it db.sh?

b-flo commented:
> I just ran your branch of yesno for now without nlsyms, but I couldn't reproduce the <blank>.
> Could you show me the token_int and token file? I'm not sure where the problem is.
> If you have time, would you investigate this problem?

The problem is that `token` contains `<blank>` when passed to `tokens2text(...)` and `text` is used for scoring: 

https://github.com/espnet/espnet/blob/e0fd073a70bcded6a0e6a3587630410a994ccdb8/espnet2/bin/asr_decode.py#L179-L193

I directly fixed though `token_id` because it was convenient and consistent in regards to other outputs but tell me if you want a different fix!

kamo-naoyuki commented on espnet2/bin/asr_decode.py:
```diff
@@ -193,16 +193,19 @@ def recog(
-                # remove sos/eos and get results
+                # remove sos/eos, blank and get results
                 token_int = hyp.yseq[1:-1].tolist()
+                token_int = list(filter(lambda x: x != 0, token_int))
```
I see, thank you!! I think this shape is good.
Would you split "remove sos/eos, blank and get results" and put a comment such that It is assumed that `<blank>` symbol equal to 0 ?

b-flo replied:
Done!

kamo-naoyuki commented:
Thanks! LGTM except for feat_type.

@sw005320 @kan-bayashi How do you think about putting "RESULTS.md" as "README.md" at under egs2/asr1/*? It can be shown when visiting, so more easy to get the result.

kan-bayashi commented:
> How do you think about putting "RESULTS.md" as "README.md" at under egs2/asr1/*? It can be shown when visiting, so more easy to get the result.

I agree. We can also add info about the recipe itself in addition to the results. (e.g. how to set database path for non-free corpus.) It is better to add template README.md.

b-flo commented:
> > How do you think about putting "RESULTS.md" as "README.md" at under egs2/asr1/*? It can be shown when visiting, so more easy to get the result.
> 
> I agree. We can also add info about the recipe itself in addition to the results. (e.g. how to set database path for non-free corpus.) It is better to add template README.md.

For how2 database setup, I added instructions in data preparation script, similarly to original recipe. I added the following in `local/data.sh` : 

```
# 1. Go to https://github.com/srvk/how2-dataset and follow instructions to download how2.                                                                   
#    For ASR, the following parts should be selected in form:                                 
#        (audio) fbank+pitch features in Kaldi scp/ark format                                  
#        (en) English text                                                                     
# 
#    First part (feats), contains directory 'fbank_pitch_181516'.                                                                
#    Second part (data) contains 'how2-300h-v1' with following directories:                   
#        how2-300h-v1                                                                          
#        |_ data/                                                                              
#            |_ val                                                                             
#            |_ train                                                                           
#            |_ dev5                                                                            
#        |_ features/                             
#                                                                                          
# 2. Copy 'fbank_pitch_181516' contents into 'how2-300h-v1/features.                       
#                                                                                          
# 3. Set up 'HOW2' in db.sh to where you put 'how-300h-v1' directory.  
```

If HOW2 is not set correctly in `db.sh` (which is expected), the following message will be displayed : 
  `"${datadir} doesn't exist. Please read instructions in local/data.sh."`

Is that okay?

kamo-naoyuki commented:
@b-flo How about write it in README.md? Anyway, I think this PR should be limited to vivos and yesno. 

Could you put the result as README.md? And change feat_type = "raw" and write the both results of raw and fbank_pitch, then let's merge this PR.
We are planning to change Kaldi's compiling to optional, so basically we can't use fbank_pitch as default even if it affects the result.

b-flo commented:
OK, done! I'll open another PR for chime5 and how2.

sw005320 commented on egs2/vivos/asr1/README.md:
```diff
@@ -0,0 +1,47 @@
+
+|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
+|---|---|---|---|---|---|---|---|---|
+|decode_testdecode_asr_model_eval.loss.best|760|7722|16.1|83.2|0.6|0.6|84.4|100.0|
```
why it is so bad?

sw005320 commented on egs2/vivos/asr1/README.md:
```diff
@@ -0,0 +1,47 @@
+|---|---|---|---|---|---|---|---|---|
+|decode_testdecode_asr_model_eval.loss.best|760|7722|16.1|83.2|0.6|0.6|84.4|100.0|
+|decode_testdecode_lm_eval.loss.best_asr_model_eval.loss.best|760|7722|61.1|38.3|0.6|0.6|39.4|92.9|
+|decode_train_devdecode_asr_model_eval.loss.best|100|927|13.9|86.0|0.1|0.2|86.3|100.0|
```
ditto

sw005320 commented:
Thanks, I think this is acceptable, but we need some notes.

sw005320 commented:
> Thanks, I think this is acceptable, but we need some notes for the degradation due to the pitch feature.
